### PR TITLE
[TG-1157] Add specialised java generic class type.

### PR DESCRIPTION
**Pull request by NlightNFotis**

### CODE REVIEW

File: src/java_bytecode/generate_java_generic_type.cpp [hunk 1]
```diff
@@ -87,17 +69,146 @@ symbolt generate_java_generic_typet::operator()(
...
-    new_tag,
-    symbol_table,
-    message_handler,
-    replacement_components);
+//
+//  generate_class_stub(
+//    new_tag,
+//    symbol_table,
+//    message_handler,
+//    replacement_components);
```
Comment (mgudemann) [inline src/java_bytecode/generate_java_generic_type.cpp:77]: should have commented-out code here

File: src/java_bytecode/generate_java_generic_type.cpp [hunk 2]
```diff
@@ -87,17 +69,146 @@ symbolt generate_java_generic_typet::operator()(
...
+
+  // inlined the generate_class_stub for now
+  {
+    java_specialised_generic_class_typet specialised_class;
+    specialised_class.set_tag(replacement_type.get_tag());
+    // NOTE: the tag absolutely has to be BasicGeneric
```
Comment (mgudemann) [inline src/java_bytecode/generate_java_generic_type.cpp:83]: what does `BasicGeneric` here mean?

File: src/java_bytecode/generate_java_generic_type.cpp [hunk 3]
```diff
@@ -87,17 +69,146 @@ symbolt generate_java_generic_typet::operator()(
...
+    optionalt<size_t> results =
+      java_generics_get_index_for_subtype(generic_class, component_identifier);
+
+    INVARIANT(results.has_value(), "generic component type not found");
+
+    if(results)
+    {
+      return generic_reference.generic_type_variables()[*results];
+    }
+    else
```
Comment (mgudemann) [inline src/java_bytecode/generate_java_generic_type.cpp:146]: with the above invariant, is this case possible at all ?

File: src/java_bytecode/generate_java_generic_type.cpp [hunk 4]
```diff
@@ -87,17 +69,146 @@ symbolt generate_java_generic_typet::operator()(
...
+              return java_generic_inst_parametert(
+                to_symbol_type(replacement_type.subtype()));
+            }
+          });
+
+      java_generic_typet new_type=generic_type;
+      new_type.generic_type_variables()=replaced_type_variables;
+      return new_type;
+    }
+    else if(parameter_type.subtype().id() == ID_symbol)
```
Comment (mgudemann) [inline src/java_bytecode/generate_java_generic_type.cpp:191]: what are the cases that can happen here? The only one seems to be arrays, please add a short documentation here.

File: src/java_bytecode/java_types.h [hunk 5]
```diff
@@ -403,4 +403,29 @@ inline const optionalt<size_t> java_generics_get_index_for_subtype(
   return std::distance(gen_types.cbegin(), iter);
 }
 
+class java_specialised_generic_class_typet : public java_class_typet
+{
+  // note the constryctor could take the components and construct it itself
```
Comment (mgudemann) [inline src/java_bytecode/java_types.h:408]: typo: constructor

File: src/java_bytecode/java_types.h [hunk 6]
```diff
@@ -403,4 +403,29 @@ inline const optionalt<size_t> java_generics_get_index_for_subtype(
...
+  // note vector of generic parameter of symbol type
+public:
+  // TODO: to be defined more appropriately.
```
Comment (mgudemann) [inline src/java_bytecode/java_types.h:411]: TODO for whom?

File: unit/java_bytecode/generate_concrete_generic_type/generate_java_generic_type.cpp [hunk 7]
```diff
@@ -179,27 +179,27 @@ SCENARIO(
...
   // to the generic class so that we can test the specialization of that generic
   // class
-  const irep_idt harness_class=
+  const irep_idt harness_class =
     "java::generic_field_array_instantiation";
 
   // We want to test that the specialized/instantiated class has it's field
   // type updated, so find the specialized class, not the generic class.
-  const irep_idt test_class=
+  const irep_idt test_class =
```
Comment (mgudemann) [inline unit/java_bytecode/generate_concrete_generic_type/generate_java_generic_type.cpp:187]: does this work with the changed implementation for arrays as generic parameters? These should now be of the form `java::array[reference]of_java::inner_type`

**Review by mgudemann — CHANGES_REQUESTED** [review]

**Comment by NlightNFotis** [comment]
This is being superseded by https://github.com/diffblue/cbmc/pull/1606. Thanks for your comments everyone.

**Pull request closed without merging**